Mixins refactoring + warnings fixes

MrFaBemol opened the pull request.

Commit:
[REF] Move some logic to a more general mixin

Commit:
[REF|NEW] carbon.line.mixin + Refactor of onsp_co2_purchase


The logic around carbon debt computation in lines is now centralized in a mixin.
Any "line model" like `account.move.line` or `purchase.order.line` can inherits it and apply some specific logic through methods overriding.

See docstrings for more infos.

Commit:
[FIX] Logs warnings fixes

dehlya commented on onsp_co2/migrations/16.0.1.2.0/pre-create_fill_columns.py:
```diff
@@ -0,0 +1,27 @@
+
+        queries.append(f"UPDATE account_move_line SET carbon_origin_name='{name}', carbon_origin_value='{value}' WHERE id={row[0]}")
+
+    cr.execute(";".join(queries))
```
What about having only one cr.execute() ? 
You could replace that part by this query and put it with the ALTER TABLE : 

```SQL
UPDATE account_move_line 
SET
    carbon_origin_name = CASE 
        WHEN POSITION('|' IN carbon_value_origin) > 0 
        THEN LEFT(carbon_value_origin, POSITION('|' IN carbon_value_origin) - 1)
        ELSE carbon_value_origin 
    END,
    carbon_origin_value = CASE 
        WHEN POSITION('|' IN carbon_value_origin) > 0 
        THEN RIGHT(carbon_value_origin, LENGTH(carbon_value_origin) - POSITION('|' IN carbon_value_origin))
        ELSE '0.0' 
END
WHERE
    carbon_value_origin IS NOT NULL 
    AND carbon_value_origin != ''; 
   ```

MrFaBemol replied:
many thanks!

dehlya commented on onsp_co2_purchase/migrations/16.0.1.1.0/pre-create_fill_columns.py:
```diff
@@ -0,0 +1,27 @@
+        """
+    )
+
+    queries = []
+    cr.execute("SELECT id, carbon_value_origin FROM purchase_order_line WHERE carbon_value_origin IS NOT NULL AND carbon_value_origin!='';")
+    for row in cr.fetchall():
+        origin_vals = row[1].split("|")
+        if len(origin_vals) == 2:
+            name = origin_vals[0]
+            value = origin_vals[1]
+        else:
+            name = row[1]
+            value = '0.0'
+
+        queries.append(f"UPDATE purchase_order_line SET carbon_origin_name='{name}', carbon_origin_value='{value}' WHERE id={row[0]}")
```
As for in onsp_co2, i think we could do it in the first execute with this:

```SQL
UPDATE purchase_order_line 
SET
    carbon_origin_name = CASE 
        WHEN POSITION('|' IN carbon_value_origin) > 0 
        THEN LEFT(carbon_value_origin, POSITION('|' IN carbon_value_origin) - 1)
        ELSE carbon_value_origin 
    END,
    carbon_origin_value = CASE 
        WHEN POSITION('|' IN carbon_value_origin) > 0 
        THEN RIGHT(carbon_value_origin, LENGTH(carbon_value_origin) - POSITION('|' IN carbon_value_origin))
        ELSE '0.0' 
    END
WHERE
    carbon_value_origin IS NOT NULL 
    AND carbon_value_origin != ''; 
```

Commit:
[IMP] migration scripts to fill origin columns refactored to SQL

Proposition from Dehlya (https://github.com/dehlya)
https://github.com/MyCityCO2/mycityco2-engine/pull/2
